### PR TITLE
Add to Elastic connector IpAddress field

### PR DESCRIPTION
**Pull request by eskabetxe**
This allow to retrieve IP fields and filter by a specific value

### CODE REVIEW

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst [hunk 1]
```diff
@@ -182,6 +182,7 @@ Elasticsearch Presto
 ``keyword``   ``VARCHAR``
 ``text``      ``VARCHAR``
 ``date``      ``TIMESTAMP``
+``ip``        ``VARCHAR``
```
Comment (findepi) [inline presto-docs/src/main/sphinx/connector/elasticsearch.rst:185]: see https://prestosql.io/docs/current/language/types.html#ipaddress

Comment (electrum) [reply presto-docs/src/main/sphinx/connector/elasticsearch.rst:185]: Agreed, it would be better to map this to `ipaddress`

Comment (eskabetxe) [reply presto-docs/src/main/sphinx/connector/elasticsearch.rst]: I change it.. but I'm unable to make this work..
the code to read the value is on IpAddressDecoder

can someone guide me?

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            output.appendNull();
+        }
+        else if (value instanceof String) {
+            String address = (String) value;
+            if (address.contains(".")) {
+                address = "::" + address;
+            }
+            BlockBuilder blockBuilder = IPADDRESS.createFixedSizeBlockBuilder(1);
+            Slice slice = Slices.wrappedBuffer(InetAddresses.forString(address).getAddress());
+            IPADDRESS.writeSlice(blockBuilder, slice);
```
Comment (findepi) [inline presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java:57]: You may want to follow `io.prestosql.type.IpAddressOperators#castFromVarcharToIpAddress` logic.

@electrum should we copy the method here? or is there a way to reuse?

Comment (eskabetxe) [reply presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java:57]: I still have the same error
`Caused by: java.lang.IllegalArgumentException: position is not valid
	at io.prestosql.spi.block.RunLengthEncodedBlock.checkReadablePosition(RunLengthEncodedBlock.java:311)
	at io.prestosql.spi.block.RunLengthEncodedBlock.isNull(RunLengthEncodedBlock.java:277)
	at io.prestosql.type.IpAddressType.getObjectValue(IpAddressType.java:121)`

File: presto-docs/src/main/sphinx/connector/elasticsearch.rst (hunk 1 above)
Comment (findepi) [reply presto-docs/src/main/sphinx/connector/elasticsearch.rst]: `IPADDRESS.writeSlice(blockBuilder, slice);` -> `IPADDRESS.writeSlice(output, slice);`

you don't need intermediate block build, see 
https://github.com/prestosql/presto/blob/677df2ced0306609d870c75b75da347c4a688e62/presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/VarbinaryDecoder.java#L37-L39

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java (hunk 2 above)
Comment (findepi) [reply presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java]: in current code, you're not writing to `output`

`IPADDRESS.writeSlice(blockBuilder, slice);` -> `IPADDRESS.writeSlice(output, slice);`

you don't need intermediate block build, see 
https://github.com/prestosql/presto/blob/677df2ced0306609d870c75b75da347c4a688e62/presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/VarbinaryDecoder.java#L37-L39

File: presto-elasticsearch/pom.xml [hunk 3]
```diff
@@ -303,20 +303,18 @@
...
             <groupId>io.prestosql</groupId>
             <artifactId>presto-client</artifactId>
-            <scope>test</scope>
         </dependency>
 
+
         <dependency>
             <groupId>io.prestosql</groupId>
             <artifactId>presto-main</artifactId>
-            <scope>test</scope>
```
Comment (findepi) [inline presto-elasticsearch/pom.xml:319]: I understand this is just temporary change.
This will need to be reverted, but let's focus on this once you get things working.

(https://github.com/prestosql/presto/pull/3347#discussion_r403839987)

Comment (eskabetxe) [reply presto-elasticsearch/pom.xml:319]: I have to do this to get IpAddressType

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java (hunk 2 above)
Comment (eskabetxe) [reply presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java]: yes its that..

File: presto-elasticsearch/pom.xml (hunk 3 above)
Comment (findepi) [reply presto-elasticsearch/pom.xml]: i added a fixup for this -- https://github.com/prestosql/presto/pull/3347/commits/701d2f57b35e991e0280e78a8d33eed0e7921e96

Comment (eskabetxe) [reply presto-elasticsearch/pom.xml]: @findepi  there are problems with that change
`
IpAddressDecoder.java:[20,25] package io.prestosql.type does not exist
ElasticsearchPageSource.java:[73,32] package io.prestosql.type does not exist
`
the problematic imports are
`
io.prestosql.type.IpAddressOperators
io.prestosql.type.IpAddressType
`

the change you add to ElasticSearchMetadata should be replicated on this two class?

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java (hunk 2 above)
Comment (electrum) [reply presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java]: We should copy that logic for now. Ideally, we would lookup the cast operator in `TypeManager`, but I don’t think that is supported yet and @dain has reservations about adding it due to conflicting with the function work.

Comment (eskabetxe) [reply presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java]: done

File: presto-elasticsearch/pom.xml (hunk 3 above)
Comment (eskabetxe) [reply presto-elasticsearch/pom.xml]: fixed

**Comment by martint** [comment]
Can you squash all the commits?

**Comment by eskabetxe** [comment]
@martint squash and rebase done

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            ipAddressType.writeSlice(output, slice);
+        }
+        else {
+            throw new PrestoException(TYPE_MISMATCH, format("Expected a string value for field '%s' of type IP: %s [%s]", path, value, value.getClass().getSimpleName()));
+        }
+    }
+
+    /**
+      *this is a copy of IpAddressOperators.castFromVarcharToIpAddress method
+      */
```
Comment (martint) [inline presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java:65]: Nit: use a regular comment, not javadoc:

```suggestion
    // This is a copy of IpAddressOperators.castFromVarcharToIpAddress method
```

File: presto-elasticsearch/pom.xml [hunk 5]
```diff
@@ -331,18 +330,6 @@
...
-            <artifactId>http-server</artifactId>
-            <scope>test</scope>
-        </dependency>
-
-        <dependency>
-            <groupId>io.airlift</groupId>
-            <artifactId>node</artifactId>
-            <scope>test</scope>
-        </dependency>
-
```
Comment (martint) [inline presto-elasticsearch/pom.xml:345]: Unrelated changes?

File: presto-elasticsearch/pom.xml [hunk 6]
```diff
@@ -355,12 +342,6 @@
             <scope>test</scope>
         </dependency>
 
-        <dependency>
-            <groupId>javax.servlet</groupId>
-            <artifactId>javax.servlet-api</artifactId>
-            <scope>test</scope>
-        </dependency>
```
Comment (martint) [inline presto-elasticsearch/pom.xml:362]: Unrelated change?

File: presto-elasticsearch/pom.xml [hunk 7]
```diff
@@ -303,7 +303,6 @@
         <dependency>
             <groupId>org.testcontainers</groupId>
             <artifactId>elasticsearch</artifactId>
-            <version>1.12.3</version>
```
Comment (martint) [inline presto-elasticsearch/pom.xml:306]: Unrelated change?

File: presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java (hunk 4 above)
Comment (eskabetxe) [reply presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java]: done

File: presto-elasticsearch/pom.xml (hunk 5 above)
Comment (eskabetxe) [reply presto-elasticsearch/pom.xml]: reverted

File: presto-elasticsearch/pom.xml (hunk 6 above)
Comment (eskabetxe) [reply presto-elasticsearch/pom.xml]: reverted

File: presto-elasticsearch/pom.xml (hunk 7 above)
Comment (eskabetxe) [reply presto-elasticsearch/pom.xml]: reverted

**Comment by martint** [comment]
Merged, thanks for your contribution!
